File: packages/snipx-ts/src/expand.ts
/** Mirror of crates/snipx-core/src/expand.rs. */

import {
  nodeSpan,
  objectDecorations,
  objectListObjects,
  statementDecorations,
  statementObjectLists,
  statementPredicates,
  statementSubject,
} from "./ast.js";
import type { Diagnostic, SourceSpan } from "./diagnostic.js";
import type { Parse } from "./parser.js";
import type { TextSpan } from "./match.js";
import { snippetValueFromNode, type SnippetValue } from "./snippet.js";
import type { SyntaxNode } from "./syntax.js";
import { isWhitespace } from "./indexMaps.js";

export type LocalScope = "sentence" | "paragraph";
export type LocalRegion = "before" | "after" | "whole";

export interface LocalSubject {
  marker: string;
  scope: LocalScope;
  region: LocalRegion;
  textSpan: boolean;
  /** UTF-16 code-unit span of the enclosing intralinea block. */
  blockSpan: SourceSpan;
}

export type Value =
  | { type: "name"; value: string }
  | { type: "predicate"; value: string }
  | { type: "string"; value: string }
  | { type: "number"; value: number }
  | { type: "invalidNumber"; source: string }
  | { type: "boolean"; value: boolean }
  | { type: "uri"; value: string }
  | { type: "snippet"; snippet: SnippetValue }
  | { type: "textSpanSnippet"; snippet: SnippetValue }
  | { type: "resolvedTextSpan"; snippet: SnippetValue; span: TextSpan }
  | { type: "localSubject"; local: LocalSubject }
  | { type: "wholeDocument" }
  | { type: "unresolved"; source: string }
  | { type: "unresolvedLocalSubject"; marker: string };

export interface ExpandedStatement {
  subject: Value;
  /** Spans in UTF-16 code units of the source. */
  subjectSpan: SourceSpan | null;
  predicate: Value;
  predicateSpan: SourceSpan | null;
  object: Value;
  objectSpan: SourceSpan | null;
  statementSpan: SourceSpan;
}

export interface ExpandOptions {
  ambientSubject: Value | null;
}

export interface ExpandResult {
  statements: ExpandedStatement[];
  diagnostics: Diagnostic[];
}

export function expand(parse: Parse, options: ExpandOptions): ExpandResult {
  const result: ExpandResult = {
    statements: [],
    diagnostics: [...parse.diagnostics],
  };

  for (const node of parse.root.descendants()) {
    if (node.kind === "Statement") {
      expandStatement(node, options, result, parse);
    }
  }

  return result;
}

function expandStatement(
  statement: SyntaxNode,
  options: ExpandOptions,
  result: ExpandResult,
  parse: Parse,
): void {
  const statementSpan = nodeSpan(statement);
  const explicitSubject = statementSubject(statement);
  let subjectSpan: SourceSpan | null =
    explicitSubject !== null ? nodeSpan(explicitSubject) : null;
  let subject: Value | null = null;
  if (explicitSubject !== null) {
    subject = valueFromNode(explicitSubject);
  }
  if (subject === null) {
    const local = localSubjectValue(statement);
    if (local !== null) {
      subjectSpan = local[1];
      subject = local[0];
    } else {
      subject = options.ambientSubject;
      if (explicitSubject === null) {
        subjectSpan = null;
      }
    }
  }

  if (subject === null) {
    result.diagnostics.push({
      code: "MissingAmbientSubject",
      severity: "error",
      message: "Subjectless statement requires an ambient subject",
      span: nodeSpan(statement),
      related: [],
    });
    return;
  }
  diagnoseInvalidNumber(subject, subjectSpan, result);

  for (const decoration of statementDecorations(statement)) {
    pushDecoration(subject, subjectSpan, decoration, result);
  }

  const predicates = statementPredicates(statement);
  const objectLists = statementObjectLists(statement);
  const pairCount = Math.min(predicates.length, objectLists.length);
  for (let i = 0; i < pairCount; i += 1) {
    const predicateNode = predicates[i];
    const objectList = objectLists[i];
    if (predicateNode === undefined || objectList === undefined) continue;
    const predicateSpan = nodeSpan(predicateNode);
    const predicate: Value = { type: "predicate", value: predicateText(predicateNode) };

    for (const object of objectListObjects(objectList)) {
      const value = valueFromNode(object);
      if (value === null) {
        continue;
      }
      const objectSpan = nodeSpan(object);
      diagnoseInvalidNumber(value, objectSpan, result);
      result.statements.push({
        subject,
        subjectSpan,
        predicate,
        predicateSpan,
        object: value,
        objectSpan,
        statementSpan,
      });

      for (const decoration of objectDecorations(object)) {
        pushDecoration(value, objectSpan, decoration, result);
      }
    }
  }
  void parse;
}

function pushDecoration(
  subject: Value,
  subjectSpan: SourceSpan | null,
  decoration: SyntaxNode,
  result: ExpandResult,
): void {
  const objectNode =
    decoration
      .childNodes()
      .find((child) => child.kind === "String" || child.kind === "TripleString") ?? null;
  const object = objectNode !== null ? valueFromNode(objectNode) : null;

  if (object === null) {
    result.diagnostics.push({
      code: "InvalidDecorationTarget",
      severity: "error",
      message: "Decoration requires a quoted string",
      span: nodeSpan(decoration),
      related: [],
    });
    return;
  }

  result.statements.push({
    subject,
    subjectSpan,
    predicate: { type: "predicate", value: "note" },
    predicateSpan: null,
    object,
    objectSpan: objectNode !== null ? nodeSpan(objectNode) : null,
    statementSpan: nodeSpan(decoration),
  });
}

function localSubjectValue(statement: SyntaxNode): [Value, SourceSpan] | null {
  let markerNode: SyntaxNode | null = null;
  for (const node of statement.descendants()) {
    if (node.kind === "LocalSubjectMarker") {
      markerNode = node;
      break;
    }
  }
  if (markerNode === null) return null;
  const marker = markerNode.toText();
  let textSpan = false;
  let body = marker;
  if (marker.startsWith("~")) {
    textSpan = true;
    body = marker.slice(1);
  }
  let scope: LocalScope;
  let region: LocalRegion;
  switch (body) {
    case "<":
      scope = "sentence";
      region = "before";
      break;
    case ">":
      scope = "sentence";
      region = "after";
      break;
    case "<>":
      scope = "sentence";
      region = "whole";
      break;
    case "<<":
      scope = "paragraph";
      region = "before";
      break;
    case ">>":
      scope = "paragraph";
      region = "after";
      break;
    case "<<>>":
      scope = "paragraph";
      region = "whole";
      break;
    default:
      return null;
  }
  let block: SyntaxNode | null = null;
  for (const ancestor of markerNode.ancestors()) {
    if (ancestor.kind === "IntralineaBlock") {
      block = ancestor;
      break;
    }
  }
  if (block === null) return null;
  const span = nodeSpan(markerNode);
  return [
    {
      type: "localSubject",
      local: {
        marker,
        scope,
        region,
        textSpan,
        blockSpan: nodeSpan(block),
      },
    },
    span,
  ];
}

function predicateText(node: SyntaxNode): string {
  const text = trimRust(node.toText());
  if (text.startsWith("`") && text.endsWith("`") && text.length >= 2) {
    return text.slice(1, -1);
  }
  return text;
}

const VALUE_KINDS = new Set([
  "Snippet",
  "RangeSnippet",
  "Uri",
  "String",
  "TripleString",
  "Number",
  "Boolean",
  "Identifier",
]);

export function valueFromNode(node: SyntaxNode): Value | null {
  let valueNode: SyntaxNode | null = null;
  for (const candidate of node.descendants()) {
    if (VALUE_KINDS.has(candidate.kind)) {
      valueNode = candidate;
      break;
    }
  }
  if (valueNode === null) return null;
  const text = valueNode.toText();

  switch (valueNode.kind) {
    case "Snippet":
    case "RangeSnippet": {
      const valueText = node.toText();
      const syntax = trimRust(valueText);
      let textSpan = false;
      let source = syntax;
      if (syntax.startsWith("~")) {
        textSpan = true;
        source = syntax.slice(1);
      }
      const snippet = snippetValueFromNode(valueNode, source);
      return textSpan
        ? { type: "textSpanSnippet", snippet }
        : { type: "snippet", snippet };
    }
    case "Uri": {
      const stripped =
        text.startsWith("<") && text.endsWith(">") && text.length >= 2
          ? text.slice(1, -1)
          : text;
      return { type: "uri", value: stripped };
    }
    case "String":
      return { type: "string", value: unescape(unquote(text, 1)) };
    case "TripleString":
      return { type: "string", value: dedent(unquote(text, 3)) };
    case "Number": {
      const number = parseRustF64(text);
      if (number === null) return null;
      return Number.isFinite(number)
        ? { type: "number", value: number }
        : { type: "invalidNumber", source: text };
    }
    case "Boolean":
      if (text === "true") return { type: "boolean", value: true };
      if (text === "false") return { type: "boolean", value: false };
      return null;
    case "Identifier":
      return { type: "name", value: text };
    default:
      return null;
  }
}

/** Rust `str::parse::<f64>` for the lexical forms parse_number emits. */
function parseRustF64(text: string): number | null {
  if (!/^-?(\d+)(\.\d+)?$/.test(text)) {
    return null;
  }
  return Number(text);
}

function diagnoseInvalidNumber(
  value: Value,
  span: SourceSpan | null,
  result: ExpandResult,
): void {
  if (value.type === "invalidNumber") {
    result.diagnostics.push({
      code: "InvalidNumber",
      severity: "error",
      message: `JSON number is outside the finite range: ${value.source}`,
      span,
      related: [],
    });
  }
}

function unescape(text: string): string {
  let output = "";
  let i = 0;
  while (i < text.length) {
    const ch = text[i];
    if (ch === undefined) break;
    if (ch !== "\\") {
      output += ch;
      i += 1;
      continue;
    }
    const next = text[i + 1];
    switch (next) {
      case "n":
        output += "\n";
        i += 2;
        break;
      case "t":
        output += "\t";
        i += 2;
        break;
      case "r":
        output += "\r";
        i += 2;
        break;
      case "\\":
        output += "\\";
        i += 2;
        break;
      case '"':
        output += '"';
        i += 2;
        break;
      case "'":
        output += "'";
        i += 2;
        break;
      case "0":
        output += "\0";
        i += 2;
        break;
      case undefined:
        output += "\\";
        i += 1;
        break;
      // Unknown escapes stay verbatim rather than silently dropping text.
      default:
        output += "\\" + next;
        i += 2;
        break;
    }
  }
  return output;
}

function dedent(text: string): string {
  let body: string;
  if (text.startsWith("\n")) {
    body = text.slice(1);
  } else if (text.startsWith("\r\n")) {
    body = text.slice(2);
  } else {
    return text;
  }

  const lines = body.split("\n");
  const indents = lines
    .filter((line) => trimRust(line).length > 0)
    .map((line) => indentLen(line));
  const commonIndent = indents.length > 0 ? Math.min(...indents) : 0;

  return lines
    .map((line) => line.slice(Math.min(commonIndent, indentLen(line))))
    .join("\n");
}

function indentLen(line: string): number {
  let i = 0;
  while (i < line.length && (line[i] === " " || line[i] === "\t")) {
    i += 1;
  }
  return i;
}

function unquote(text: string, quoteWidth: number): string {
  if (text.length < 2 * quoteWidth) return text;
  return text.slice(quoteWidth, text.length - quoteWidth);
}

/** Rust `str::trim`: Unicode White_Space from both ends. */
export function trimRust(text: string): string {
  let start = 0;
  while (start < text.length) {
    const cp = text.codePointAt(start);
    if (cp === undefined) break;
    const ch = String.fromCodePoint(cp);
    if (!isWhitespace(ch)) break;
    start += ch.length;
  }
  let end = text.length;
  while (end > start) {
    const code = text.charCodeAt(end - 1);
    const prev = code >= 0xdc00 && code <= 0xdfff && end >= 2 ? end - 2 : end - 1;
    const cp = text.codePointAt(prev);
    if (cp === undefined) break;
    if (!isWhitespace(String.fromCodePoint(cp))) break;
    end = prev;
  }
  return text.slice(start, end);
}

File: crates/snipx-core/src/expand.rs
use crate::ast::{AstNode, Decoration, Statement};
use crate::diagnostic::{Diagnostic, DiagnosticCode, Severity, SourceSpan};
use crate::parser::Parse;
use crate::r#match::TextSpan;
use crate::snippet::SnippetValue;
use crate::syntax::{SyntaxKind, SyntaxNode};

#[derive(Debug, Clone, PartialEq)]
pub enum Value {
    Name(String),
    Predicate(String),
    String(String),
    Number(f64),
    InvalidNumber(String),
    Boolean(bool),
    Uri(String),
    Snippet(SnippetValue),
    TextSpanSnippet(SnippetValue),
    /// A text-span snippet after resolution, pinned to one concrete
    /// matched span. Produced only by resolve, never by expand.
    ResolvedTextSpan {
        snippet: SnippetValue,
        span: TextSpan,
    },
    LocalSubject(LocalSubject),
    WholeDocument,
    Unresolved(String),
    UnresolvedLocalSubject(String),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LocalScope {
    Sentence,
    Paragraph,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LocalRegion {
    Before,
    After,
    Whole,
}

/// An intralinea local subject marker (`<`, `>`, `<>`, `<<`, `>>`,
/// `<<>>`, optionally `~`-prefixed), anchored to its enclosing
/// `{{ ... }}` block for resolution against the stripped visible text.
#[derive(Debug, Clone, PartialEq)]
pub struct LocalSubject {
    pub marker: String,
    pub scope: LocalScope,
    pub region: LocalRegion,
    pub text_span: bool,
    pub block_span: SourceSpan,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ExpandedStatement {
    pub subject: Value,
    pub subject_span: Option<SourceSpan>,
    pub predicate: Value,
    pub predicate_span: Option<SourceSpan>,
    pub object: Value,
    pub object_span: Option<SourceSpan>,
    pub statement_span: SourceSpan,
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct ExpandOptions {
    pub ambient_subject: Option<Value>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ExpandResult {
    pub statements: Vec<ExpandedStatement>,
    pub diagnostics: Vec<Diagnostic>,
}

pub fn expand(parse: &Parse, options: ExpandOptions) -> ExpandResult {
    let mut result = ExpandResult {
        statements: Vec::new(),
        diagnostics: parse.diagnostics().to_vec(),
    };

    for node in parse
        .syntax()
        .descendants()
        .filter(|node| node.kind() == SyntaxKind::Statement)
    {
        if let Some(statement) = Statement::cast(node) {
            expand_statement(&statement, &options, &mut result);
        }
    }

    result
}

fn expand_statement(statement: &Statement, options: &ExpandOptions, result: &mut ExpandResult) {
    let statement_span = source_span(statement.syntax());
    let explicit_subject = statement.subject();
    let mut subject_span = explicit_subject
        .as_ref()
        .map(|subject| source_span(subject.syntax()));
    let subject = match explicit_subject.and_then(|subject| value_from_node(subject.syntax())) {
        Some(value) => Some(value),
        None => match local_subject_value(statement) {
            Some((value, span)) => {
                subject_span = Some(span);
                Some(value)
            }
            None => options.ambient_subject.clone(),
        },
    };

    let Some(subject) = subject else {
        result.diagnostics.push(Diagnostic {
            code: DiagnosticCode::MissingAmbientSubject,
            severity: Severity::Error,
            message: "Subjectless statement requires an ambient subject".to_owned(),
            span: Some(source_span(statement.syntax())),
            related: Vec::new(),
        });
        return;
    };
    diagnose_invalid_number(&subject, subject_span.clone(), result);

    for decoration in statement.decorations() {
        push_decoration(&subject, subject_span.clone(), &decoration, result);
    }

    for (predicate, object_list) in statement.predicates().zip(statement.object_lists()) {
        let predicate_span = Some(source_span(predicate.syntax()));
        let predicate = Value::Predicate(predicate_text(predicate.syntax()));

        for object in object_list.objects() {
            let Some(value) = value_from_node(object.syntax()) else {
                continue;
            };
            let object_span = Some(source_span(object.syntax()));
            diagnose_invalid_number(&value, object_span.clone(), result);
            result.statements.push(ExpandedStatement {
                subject: subject.clone(),
                subject_span: subject_span.clone(),
                predicate: predicate.clone(),
                predicate_span: predicate_span.clone(),
                object: value.clone(),
                object_span: object_span.clone(),
                statement_span: statement_span.clone(),
            });

            for decoration in object.decorations() {
                push_decoration(&value, object_span.clone(), &decoration, result);
            }
        }
    }
}

fn push_decoration(
    subject: &Value,
    subject_span: Option<SourceSpan>,
    decoration: &Decoration,
    result: &mut ExpandResult,
) {
    let object_node = decoration
        .syntax()
        .children()
        .find(|child| matches!(child.kind(), SyntaxKind::String | SyntaxKind::TripleString));
    let object = object_node.as_ref().and_then(value_from_node);

    let Some(object) = object else {
        result.diagnostics.push(Diagnostic {
            code: DiagnosticCode::InvalidDecorationTarget,
            severity: Severity::Error,
            message: "Decoration requires a quoted string".to_owned(),
            span: Some(source_span(decoration.syntax())),
            related: Vec::new(),
        });
        return;
    };

    result.statements.push(ExpandedStatement {
        subject: subject.clone(),
        subject_span,
        predicate: Value::Predicate("note".to_owned()),
        predicate_span: None,
        object,
        object_span: object_node.as_ref().map(source_span),
        statement_span: source_span(decoration.syntax()),
    });
}

fn local_subject_value(statement: &Statement) -> Option<(Value, SourceSpan)> {
    let marker_node = statement
        .syntax()
        .descendants()
        .find(|node| node.kind() == SyntaxKind::LocalSubjectMarker)?;
    let marker = marker_node.to_string();
    let (text_span, body) = match marker.strip_prefix('~') {
        Some(rest) => (true, rest),
        None => (false, marker.as_str()),
    };
    let (scope, region) = match body {
        "<" => (LocalScope::Sentence, LocalRegion::Before),
        ">" => (LocalScope::Sentence, LocalRegion::After),
        "<>" => (LocalScope::Sentence, LocalRegion::Whole),
        "<<" => (LocalScope::Paragraph, LocalRegion::Before),
        ">>" => (LocalScope::Paragraph, LocalRegion::After),
        "<<>>" => (LocalScope::Paragraph, LocalRegion::Whole),
        _ => return None,
    };
    let block = marker_node
        .ancestors()
        .find(|node| node.kind() == SyntaxKind::IntralineaBlock)?;
    let span = source_span(&marker_node);
    Some((
        Value::LocalSubject(LocalSubject {
            marker,
            scope,
            region,
            text_span,
            block_span: source_span(&block),
        }),
        span,
    ))
}

fn predicate_text(node: &SyntaxNode) -> String {
    let text = node.to_string();
    let trimmed = text.trim();
    trimmed
        .strip_prefix('`')
        .and_then(|text| text.strip_suffix('`'))
        .unwrap_or(trimmed)
        .to_owned()
}

fn value_from_node(node: &SyntaxNode) -> Option<Value> {
    let value_node = node.descendants().find(|candidate| {
        matches!(
            candidate.kind(),
            SyntaxKind::Snippet
                | SyntaxKind::RangeSnippet
                | SyntaxKind::Uri
                | SyntaxKind::String
                | SyntaxKind::TripleString
                | SyntaxKind::Number
                | SyntaxKind::Boolean
                | SyntaxKind::Identifier
        )
    })?;
    let text = value_node.to_string();

    match value_node.kind() {
        SyntaxKind::Snippet | SyntaxKind::RangeSnippet => {
            let value_text = node.to_string();
            let syntax = value_text.trim();
            let (text_span, source) = match syntax.strip_prefix('~') {
                Some(rest) => (true, rest),
                None => (false, syntax),
            };
            let snippet = SnippetValue::from_node(&value_node, source.to_owned());
            Some(if text_span {
                Value::TextSpanSnippet(snippet)
            } else {
                Value::Snippet(snippet)
            })
        }
        SyntaxKind::Uri => Some(Value::Uri(
            text.strip_prefix('<')
                .and_then(|text| text.strip_suffix('>'))
                .unwrap_or(&text)
                .to_owned(),
        )),
        SyntaxKind::String => Some(Value::String(unescape(&unquote(&text, 1)))),
        SyntaxKind::TripleString => Some(Value::String(dedent(&unquote(&text, 3)))),
        SyntaxKind::Number => text.parse().ok().map(|number: f64| {
            if number.is_finite() {
                Value::Number(number)
            } else {
                Value::InvalidNumber(text)
            }
        }),
        SyntaxKind::Boolean => match text.as_str() {
            "true" => Some(Value::Boolean(true)),
            "false" => Some(Value::Boolean(false)),
            _ => None,
        },
        SyntaxKind::Identifier => Some(Value::Name(text)),
        _ => None,
    }
}

fn diagnose_invalid_number(value: &Value, span: Option<SourceSpan>, result: &mut ExpandResult) {
    if let Value::InvalidNumber(source) = value {
        result.diagnostics.push(Diagnostic {
            code: DiagnosticCode::InvalidNumber,
            severity: Severity::Error,
            message: format!("JSON number is outside the finite range: {source}"),
            span,
            related: Vec::new(),
        });
    }
}

fn unescape(text: &str) -> String {
    let mut output = String::with_capacity(text.len());
    let mut chars = text.chars();
    while let Some(ch) = chars.next() {
        if ch != '\\' {
            output.push(ch);
            continue;
        }
        match chars.next() {
            Some('n') => output.push('\n'),
            Some('t') => output.push('\t'),
            Some('r') => output.push('\r'),
            Some('\\') => output.push('\\'),
            Some('"') => output.push('"'),
            Some('\'') => output.push('\''),
            Some('0') => output.push('\0'),
            // Unknown escapes stay verbatim rather than silently dropping text.
            Some(other) => {
                output.push('\\');
                output.push(other);
            }
            None => output.push('\\'),
        }
    }
    output
}

fn dedent(text: &str) -> String {
    let Some(body) = text
        .strip_prefix('\n')
        .or_else(|| text.strip_prefix("\r\n"))
    else {
        return text.to_owned();
    };

    let lines: Vec<&str> = body.split('\n').collect();
    let common_indent = lines
        .iter()
        .filter(|line| !line.trim().is_empty())
        .map(|line| line.len() - line.trim_start_matches([' ', '\t']).len())
        .min()
        .unwrap_or(0);

    lines
        .iter()
        .map(|line| {
            line.get(common_indent.min(indent_len(line))..)
                .unwrap_or("")
        })
        .collect::<Vec<_>>()
        .join("\n")
}

fn indent_len(line: &str) -> usize {
    line.len() - line.trim_start_matches([' ', '\t']).len()
}

fn unquote(text: &str, quote_width: usize) -> String {
    text.get(quote_width..text.len().saturating_sub(quote_width))
        .unwrap_or(text)
        .to_owned()
}

fn source_span(node: &SyntaxNode) -> SourceSpan {
    let range = node.text_range();
    SourceSpan {
        start: u32::from(range.start()) as usize,
        end: u32::from(range.end()) as usize,
    }
}
